### PR TITLE
Move libfaasm to this repo

### PR DESCRIPTION
**Pull request by Shillaker**
This is the first step towards #11 

The Faasm C/C++ library (`libfaasm`) defines the Faasm host interface, hence is essential for building C/C++ applications for Faasm (as well as building Faasm itself). The library currently lives in the Faasm repo, so to build such applications, you need both this toolchain _and_ the Faasm repo. By moving the library in here, we remove this extra dependency on Faasm. 

This is particularly useful for the faasm CPython build which currently has a nasty circular dependency with libfaasm and Faasm.

The more complex parts of `libfaasm` dealing with matrices in state are really just specific to the SGD experiment, so once that has a repo to itself, I'll move them out of here and into there.

### CODE REVIEW

File: bin/here.py [hunk 1]
```diff
@@ -20,7 +20,7 @@ def main():
         "-v {}:/work".format(cwd),
         "-w /work",
         "-it",
-        "faasm/sysroot:v{}".format(toolchain_ver),
+        "faasm/sysroot:{}".format(toolchain_ver),
```
Comment (Shillaker) [inline bin/here.py:23]: Containers in Docker Hub used to have a `v` before the version, but this is fixed now

File: libfaasm/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,76 @@
+cmake_minimum_required(VERSION 3.0)
```
Comment (Shillaker) [inline libfaasm/CMakeLists.txt:1]: This is just copied from Faasm

File: libfaasm/faasm/core.h [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#define BYTES(arr) reinterpret_cast<uint8_t*>(arr)
+
+    /**
```
Comment (csegarragonz) [inline libfaasm/faasm/core.h:15]: How's that this declarations are indented?

**Review by csegarragonz — APPROVED** [review]

File: libfaasm/faasm/core.h (hunk 3 above)
Comment (Shillaker) [reply libfaasm/faasm/core.h:15]: Hmm, good spot, I'll add clang-format etc. to this repo now that we have our own code in here

**Pull request merged**